crates/task-impls/src/quorum_proposal.rs

```diff
@@ -0,0 +1,477 @@
+                        HotShotEvent::ViewSyncFinalizeCertificate2Recv(view_sync_cert) => {
+                            view_sync_cert.view_number
+                        }
+                        HotShotEvent::QCFormed(cert) => match cert {
+                            either::Right(timeout) => timeout.view_number,
+                            either::Left(qc) => qc.view_number,
+                        },
+                        HotShotEvent::QuorumProposalRecv(proposal, _) => proposal.data.view_number,
+                        _ => return false,
+                    },
+                    ProposalDependency::PayloadAndMetadata => {
+                        if let HotShotEvent::SendPayloadCommitmentAndMetadata(
+                            _payload_commitment,
+                            _metadata,
+                            view,
+                        ) = event
+                        {
+                            *view
+                        } else {
+                            return false;
+                        }
+                    }
+                };
+                event_view == view_number
+            }),
+        )
+    }
+
+    /// Create and store an [`AndDependency`] combining [`EventDependency`]s associated with the
+    /// given view number if it doesn't exist. Also takes in the received `event` to seed a
+    /// dependency as already completed. This allows for the task to receive a proposable event
+    /// without losing the data that it received, as the dependency task would otherwise have no
+    /// ability to receive the event and, thus, would never propose.
+    fn create_dependency_task_if_new(
+        &mut self,
+        view_number: TYPES::Time,
+        event_receiver: Receiver<Arc<HotShotEvent<TYPES>>>,
+        event_sender: Sender<Arc<HotShotEvent<TYPES>>>,
+        event: Arc<HotShotEvent<TYPES>>,
+    ) {
+        debug!("Attempting to make dependency task for event {:?}", event);
+        if self.propose_dependencies.get(&view_number).is_some() {
+            debug!("Task already exists");
+            return;
+        }
+
+        let mut proposal_cert_validated_dependency = self.create_event_dependency(
+            ProposalDependency::ProposalCertificate,
+            view_number,
+            event_receiver.clone(),
+        );
+
+        let mut payload_commitment_dependency = self.create_event_dependency(
+            ProposalDependency::PayloadAndMetadata,
+            view_number,
+            event_receiver,
+        );
+
+        match event.as_ref() {
+            HotShotEvent::SendPayloadCommitmentAndMetadata(_, _, _) => {
+                payload_commitment_dependency.mark_as_completed(event.clone());
+            }
+            HotShotEvent::QuorumProposalRecv(_, _)
+            | HotShotEvent::QCFormed(_)
+            | HotShotEvent::ViewSyncFinalizeCertificate2Recv(_) => {
+                proposal_cert_validated_dependency.mark_as_completed(event);
+            }
```

Comment:
I don't think these 3 things should be grouped together.  The Proposal is a separate event.

For example, if we are leader for view `n`, to propose we'd need both a QC and Proposal from view `n-1` OR a ViewSyncFinalizedCertificate or TimeoutCertificate